Extend MemoryEffects to Support Target-Specific Memory Locations

CarolineConcatto opened the pull request:
This patch introduces preliminary support for additional memory locations. 
They are: target_mem0 and target_mem1 and they model  memory locations that cannot be represented with existing memory locations.

It was a solution suggested in : https://discourse.llvm.org/t/rfc-improving-fpmr-handling-for-fp8-intrinsics-in-llvm/86868/6

Currently, these locations are not yet target-specific. The goal is to enable the compiler to express read/write effects on these resources.

CarolineConcatto committed:
[RFC][Draft] Extend MemoryEffects to Support Target-Specific Memory Locations

This patch introduces preliminary support for additional memory locations,
such as FPMR and ZA, needed to model AArch64 architectural registers as
memory dependencies.

Currently, these locations are not yet target-specific. The goal is to enable
the compiler to express read/write effects on these resources.

What This Patch Does:
  Adds two new memory locations: FPMR and ZA, intended to represent
AArch64-specific inaccessible memory types.

Current Limitations:
  These new locations are not yet target-specific in the type-safe sense,
they are globally visible and hardcoded.
  There is no mechanism yet to associate a memory location with its
corresponding target (e.g., AArch64 vs RISCV).
  No changes are made yet to bitcode serialization, parser support, or alias
analysis behavior.

This patch is not functionally complete — it is a structural prototype to
solicit feedback on the direction and I would like some suggestion on
how to proceed.

CarolineConcatto commented on llvm/include/llvm/Support/ModRef.h:
```diff
@@ -56,6 +56,11 @@ enum class ModRefInfo : uint8_t {
 /// Debug print ModRefInfo.
 LLVM_ABI raw_ostream &operator<<(raw_ostream &OS, ModRefInfo MR);
 
+enum class InaccessibleTargetMemLocation {
```
Each Target should have an InaccessibleTargetMemLocation, but I am not sure if we should check which Target it is being used in Support, Attributes and so on. I  would appreciate some suggestion if this is the right way to do this.

And this is still using Data to model the Memory Locations with ModRef. Is this the only way to add new memory location?

CarolineConcatto replied:
Hi @antoniofrighetto,
I think your suggestion need me to change the way Memory Location is implemented.
However I did follow @nikic's solution(or I think I did) and create a generic name for the target memory locations, and they are printed according to Triple.  I think this makes It generic enough. 
Let me know what you think about the latest change. 
I have some follow up patches :
https://github.com/llvm/llvm-project/pull/154141
https://github.com/llvm/llvm-project/pull/154144
https://github.com/llvm/llvm-project/pull/155590
that I will rebase once this patch is merged.

CarolineConcatto commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -656,6 +660,15 @@ std::string Attribute::getAsString(bool InAttrGrp) const {
       case IRMemLocation::Other:
         llvm_unreachable("This is represented as the default access kind");
+      default: {
+        InaccessibleTargetMemLocation TargetLoc =
```
How can we make this to not be target specific? Any suggestion?

antoniofrighetto commented on llvm/lib/AsmParser/LLLexer.cpp:
```diff
@@ -701,10 +701,14 @@ lltok::Kind LLLexer::LexIdentifier() {
   KEYWORD(write);
   KEYWORD(readwrite);
   KEYWORD(argmem);
+  KEYWORD(fpmr);
+  KEYWORD(za);
```
These should be aarch64_fpmr and aarch64_za, right?

nikic commented:
Making things target-specific at this level is not worthwhile. Similar to how intrinsics are always defined, regardless of whether the corresponding target was built, these can also just always be available, but will only be used for specific targets.

nikic commented on llvm/test/Bitcode/attributes.ll:
```diff
@@ -567,6 +568,42 @@ define void @captures(ptr captures(address) %p) {
 }
 
+; CHECK: define void @fpmr_write() [[ATTR59:#.*]] {
+define void @fpmr_write() inaccessiblewritememonly(fpmr) {
```
Correct

CarolineConcatto committed:
Address review comments about tests and names

This patch is removing the keyworkd and token inaccessibleWrite and
inaccessibleRead. It is using Read and Write to set the Target Memory
Location.

Adding "aarch64" in front of the target specific memory locations

CarolineConcatto committed:
Remove old function header that doesn't exist

CarolineConcatto commented on llvm/test/TableGen/intrinsic-attrs-fp8.td:
```diff
@@ -0,0 +1,54 @@
+// CHECK-NEXT:      Attribute::get(C, Attribute::NoSync),
+// CHECK-NEXT:      Attribute::get(C, Attribute::NoFree),
+// CHECK-NEXT:      Attribute::get(C, Attribute::WillReturn),
+// CHECK-NEXT:      // ArgMem: ModRef, InaccessibleMem: ModRef, ErrnoMem: ModRef, AARCH64_FPMR: Ref, AARCH64_ZA: Mod, Other: ModRef
```
I've just noticed that this should be ArgMem: NoModRef, InaccessibleMem: NoModRef,  ErrnoMem: NoModRef.
I will see how to fix that in CodeGenIntrinsics.cpp and being able to have more than 1 location with ModRef

CarolineConcatto committed:
Fix test when target memory locations are combined

CarolineConcatto committed:
Fix codegen for opencl tests

CarolineConcatto commented:
I am removing from draft, but we are still open to suggestions. Maybe removing from draft to get more attention

jurahul commented on llvm/lib/TableGen/Record.cpp:
```diff
@@ -3102,6 +3102,21 @@ Record::getValueAsListOfDefs(StringRef FieldName) const {
   return Defs;
 }
 
+llvm::IRMemLocation Record::getLocationTypeAsInt(StringRef FieldName) const {
```
This function does not belong here. It should be moved instead to the specific backend that uses it (intrinsic emitter I am assuming)

CarolineConcatto replied:
Hello @jurahul 
Do you mind expanding what you would like to see?
Because this function is called  by CodeGenIntrinsic::setProperty(const Record *R)  in CodeGenIntrinsics.cpp, in there , AFAIK I cannot check the target architecture. So I would like some pointer to know how I should add this in the backend  of the target specific architecture.

CarolineConcatto replied:
Thank you for clarifying.
I believe I've addressed your comment in the latest update.

paulwalker-arm commented on llvm/include/llvm/AsmParser/LLToken.h:
```diff
@@ -202,6 +202,8 @@ enum Kind {
   kw_readwrite,
   kw_argmem,
   kw_inaccessiblemem,
+  kw_aarch64_fpmr,
```
Is this new? There are many examples mainly linked to calling conventions are function/argument attributes.

nikic commented:
> Hi @nikic ,
> 
> Thank you for your comments, I do have some extra notes.
> 
> My concern is that with Data being 32 bits and reserving 2 bits for each location, there are 16 memory locations available. There are 3 already being reserved to ArgMem, InaccessibleMem and Other. Now AArch64 is using 2(AArch64_FPRM and AArch64_ZA, I will probably expand to use for ZT0), if each target adds its own location it could be out of space with Data.
> 
> Any suggestion how to implement the Target Memory Location? The way it is implemented the Target Memory Locations for now is correct ? Should I change?

We could easily increase the size to 64 bits, which gives us 32 locations.

My primary concern with making this dependent on the backend is that I don't want to thread backend information through to every place that works with MemoryEffects.

However, I guess that generic code doesn't actually care about what a target-specific memory location refers to. So a possible approach would be to just reserve a fixed number of target-memory locations like `target1`, `target2`, `target3` etc. and then have the backend assign specific meanings to those locations. This can involve some support to register names for these memory locations, so they can be printed in a friendlier way. E.g. we could print them as `memory(target1: /*aarch64.fpmr*/ read)` or something, if the target information is available.

Does this sound reasonable to you?
 
> Another question, should we treat these extra locations as Inaccessible memory as well? So, should onlyAccessesInaccessibleMem() return true is any Target Memory Location is set.

That would make sense to me.

CarolineConcatto commented:
Hi @nikic ,
So your proposal will be to change this 
enum class InaccessibleTargetMemLocation {
  AARCH64_FPMR = 3,
  AARCH64_ZA = 4,
};
To be something like
enum class InaccessibleTargetMemLocation {
  TargetMemLoc1= 3,
  TargetMemLoc2 = 4,
  ....
};

The printing of _**memory(target1: /*aarch64.fpmr*/ read)**_ is happening in the file Attributes.cpp  and that is the part I may lack understanding of how to link this with the target architecture. I will need to add a Target hook in that part of the code, but atm there is not information about the target.

nikic commented:
> > However, I guess that generic code doesn't actually care about what a target-specific memory location refers to. So a possible approach would be to just reserve a fixed number of target-memory locations like target1, target2, target3 etc. and then have the backend assign specific meanings to those locations. This can involve some support to register names for these memory locations, so they can be printed in a friendlier way. E.g. we could print them as memory(target1: /_aarch64.fpmr_/ read) or something, if the target information is available.
> 
> I think one of the advantage of having out-of-line target-memory locations instead of a fixed target1..targetN bitset is that we could print them friendlier, e.g., `memory(targetmem("aarch64.za": read, "aarch64.fpmr": read))`.

But how would you represent this without making MemoryEffects an expensive data structure?

nikic commented:
> The printing of _**memory(target1: /_aarch64.fpmr_/ read)**_ is happening in the file Attributes.cpp and that is the part I may lack understanding of how to link this with the target architecture. I will need to add a Target hook in that part of the code, but atm there is not information about the target.

Right, that's the challenge with making any of this target-dependent. The best idea I have to do this without a lot of changes would be to optionally pass down the Triple to `Attribute::getAsString` and make the meaning triple-based, without having to involve the actual backend. (The mapping could be stored in TargetParser, which is where various target-specific but backend-independent information lives. And I think TableGen also links to TargetParser, which is the other place that would need this information.)

CarolineConcatto committed:
Address review comments

CarolineConcatto committed:
Remove unwanted changes from the test memory-attribute.ll

CarolineConcatto committed:
Make target memory locations generic; resolve meaning via Triple

This patch replaces target‑specific memory location names in IR
printing/parsing with generic target memory kinds whose meaning is determined
by the module’s target triple.

Replace the target specific names by generic target memory kinds
(e.g. mem_target_0, mem_target_1), and make their user‑facing names
triple‑dependent.
Plumb the Triple into attribute so textual IR reflects the correct per‑target
 names.
To print a meaningful name each target needs to add a keyword->kind
mapping so textual forms remain readable and round‑trip correctly.

CarolineConcatto commented:
Hi @nikic,
I've replaces the aarch64 target specifc memory location  names by  a generic Target memory location, and added the plumbing in AsmWriter to pass Triple to give meaning to this generic names.

CarolineConcatto committed:
Change ModRef.cpp to use generic memory target name

CarolineConcatto committed:
Fix clang format

CarolineConcatto committed:
Remove Target Triple interface for Target Memory

paulwalker-arm commented on llvm/include/llvm/AsmParser/LLToken.h:
```diff
@@ -203,6 +203,8 @@ enum Kind {
   kw_readwrite,
   kw_argmem,
   kw_inaccessiblemem,
+  kw_aarch64_fpmr,
+  kw_aarch64_za,
```
I had in mind that these would be part of the follow-on PR as well?  The idea being this PR keeps everything generic and fully supports `target_mem#` (i.e. both printing and parsing).

CarolineConcatto replied:
We need a token to be used when describing the memory:
 memory(aarch64_za : read)

There are 3 solutions:
1) Create a target specific token for each architecture(arch64_za), but without a generic token(target_mem0).  It is the one I chosen because each target needs to describe its token  to be able to use the target location . All tokens should have the architecture name, like suggested. So it could not borrow from other architectures.
So this option allows:
 memory(aarch64_za : read)
but does not allow:
 memory(target_mem0 : read)

2) Create a generic token and print what is this meaning when we add target triple support, but do not add a target specific token. There would be the tokens target_memX and when we add the triple support it could print whatever the target wants. So only the Triple.h would know the name of the target locations, not even in the description of the memory we could use a target specific name. 
So this option allows:
 memory(target_mem0 : read)
but does not allow:
 memory(aarch64_za : read)

3) Could add target specific and generic tokens. I did not liked this option because it may cause confusion  being able to describe the same thing in two ways so thing could get messy.
So this option allows:
 memory(aarch64_za : read)
and
 memory(target_mem0 : read)

Do you have a preferred solution? Or do you see another way of doing this?

CarolineConcatto committed:
Remove all Target specific tokens and definitions

It also fix this warning:
ModRef.cpp:52:5: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]

paulwalker-arm commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -656,7 +660,23 @@ std::string Attribute::getAsString(bool InAttrGrp) const {
         llvm_unreachable("This is represented as the default access kind");
+        break;
       }
+      // Target Memory locations are not IRMemLocation.
+      // It is breaking buildbots when it treats warning as error
+      if (static_cast<int>(Loc) > static_cast<int>(IRMemLocation::ErrnoMem)) {
```
Presumably this all disappears once `InaccessibleTargetMemLocation` is removed and you can just update the `switch (Loc)` block directly.

paulwalker-arm commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -640,6 +640,10 @@ std::string Attribute::getAsString(bool InAttrGrp) const {
         continue;
 
+      // Dont want to print Target Location if NoModRef
+      if (ME.isTargetMemLoc(Loc) && (MR == ModRefInfo::NoModRef))
+        continue;
```
I don't understand why this is needed. Can you elaborate?

CarolineConcatto committed:
Address review comments

CarolineConcatto committed:
Address comments in FunctionAttrs.cpp

CarolineConcatto commented on llvm/lib/Transforms/IPO/FunctionAttrs.cpp:
```diff
@@ -142,6 +142,10 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
   ME |= MemoryEffects(IRMemLocation::Other, MR);
+  // Should also set the other Target Memory Locations as MR.
+  // To compares with MemoryEffects::unknown() in addMemoryAttrs
+  ME |= MemoryEffects(IRMemLocation::TargetMem0, MR);
```
The problem I have is that the IR that is failing is a volatile:
```
 define dso_local spir_func void @non_convfun() local_unnamed_addr #0 {
entry:
  store volatile i32 0, ptr undef, align 4, !tbaa !3
  ret void
}
```
In the code in checkFunctionMemoryAccess sets the MemoryEffects  if it is volatile:
```
    // Volatile operations may access inaccessible memory.
    if (I.isVolatile())
      ME |= MemoryEffects::inaccessibleMemOnly(MR);
```
And later in addMemoryAttrs it compares with `ME == MemoryEffects::unknown() `which in the past was true because of all memory location in set to ModRef.
Now we don't set all the locations to ModRef(the targets are missing) and because of that the functions starts to identify as memory(readwrite) for the OpenCL test (convergent.cl) Which is not the case.
However is not always that we should setup the volatile target to MR
I believe we can set here because it is where we are setting ErrorMem and also Others. I thought that here is where it makes much sense because it is where we also set  Other and ErronoMem.

 Or update the test for  the 
    if (ME == MemoryEffects::unknown())
To take into consideration that Target Memory Location will never be tagged in this situation

paulwalker-arm reviewed (commented):
Sorry Carol, I didn't get chance to take a deeper look yet but here's some comments (mostly stylistic) that I had stashed.  I'll take a proper look on Monday.

CarolineConcatto committed:
Address review comments

CarolineConcatto committed:
Address comments about tablegen

paulwalker-arm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -49,6 +49,22 @@ def IntrArgMemOnly : IntrinsicProperty;
+
+
+class IntrinsicMemoryLocation;
+// These are used to represent Generic Memory Location. 
```
yes

CarolineConcatto commented on llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp:
```diff
@@ -374,7 +375,21 @@ void CodeGenIntrinsic::setProperty(const Record *R) {
+
+  } else if (R->isSubClassOf("IntrWrite")) {
+    MemoryEffects WriteMask = MemoryEffects::readOnly();
+    for (const Record *RLoc : R->getValueAsListOfDefs("Loc"))
```
It was for Record, but I can change

CarolineConcatto committed:
Address review comments

CarolineConcatto committed:
Remove unwanted changes

CarolineConcatto committed:
Set target_mem when they are not set  with inaccessiblemem

CarolineConcatto commented on llvm/lib/AsmParser/LLParser.cpp:
@paulwalker-arm  let me know if that is the behaviour we would like for the target_mem.
If so then we I believe we will need to change as well for MLIR.
This test is failing:
mlir/test/Target/LLVMIR/llvmir.mlir
I was looking at this commit:

```
commit 9b9cfe77a50abccc4b82a497e17566a454b699bd
Author: Christian Ulmann <christian.ulmann@nextsilicon.com>
Date:   Thu Jan 12 16:48:19 2023 +0100
 
     [mlir][LLVM] Replace readnone with memory effects]
```
MLIR has its own attributes, a mirror from llvm-ir,  and it maps to memory effects.

paulwalker-arm commented on llvm/include/llvm/Support/ModRef.h:
```diff
@@ -232,7 +243,10 @@ template <typename LocationEnum> class MemoryEffectsBase {
   bool onlyAccessesInaccessibleMem() const {
-    return getWithoutLoc(Location::InaccessibleMem).doesNotAccessMemory();
+    return getWithoutLoc(IRMemLocation::TargetMem0)
+        .getWithoutLoc(IRMemLocation::TargetMem1)
```
To be consistent `onlyAccessesInaccessibleOrArgMem` also needs the same additions.

paulwalker-arm commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -630,14 +630,24 @@ std::string Attribute::getAsString(bool InAttrGrp) const {
     for (auto Loc : MemoryEffects::locations()) {
       ModRefInfo MR = ME.getModRef(Loc);
-      if (MR == OtherMR)
+      // Skip target memory location
```
Perhaps:
```
// Do not print accesses that matches the default
// NOTE: Target memory locations take their default from InaccessibleMem.
```

paulwalker-arm reviewed (approved):
A final request but otherwise this looks good to me.

paulwalker-arm commented:
@nikic Do you think it'll be reasonable to revert the changes to onlyAccessesInaccessibleMem and onlyAccessesInaccessibleOrArgMem and land the patch as is? That will temporarily break the link and make it easier to rollout the support for target_mem.  Perhaps also adding a note to the LangRef saying the target_mem locations must not be used outside of internal testing?

CarolineConcatto committed:
Remove dependency between InaccessibleMem and TargetMem

Also adds a not in LangRef saying that target memories are in
development stage

CarolineConcatto committed:
Run clang-format-18

paulwalker-arm reviewed (approved):
Thanks for the updates @CarolineConcatto.
